texlive: overrideTeXConfig/withPackages

xworld21 opened the pull request:
## Description of changes
Full implementation of the multi-output packages for TeX Live and new `overrideTeXConfig`, `withPackages` functions to create combined environment. ~~, with a compatibility layer set up to help creating a scope later on (basically I copied @apfelkuchen6's `compatFixups`), and the most minimal implementation of `texlive.__buildEnv`, focussed on preserving `texlive.combine` down to the build output. The alias `texlive.__withPackages` is also in place.~~

This is now fixed to reproduce `texlive.combine` exactly, so that we have no rebuilds at this stage. What's left is to create a scope, and we'll be able to talk user visible changes.

## Things done
- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

xworld21 commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
I figured that the only realistic way to preserve `texlive.combine` to the letter is to keep its dependency resolution and filtering here, rather than trying to emulate it in `texlive.buildEnv`. Since `texlive.combine` should be deprecated after stabilizing `texlive.buildEnv`, I think it's the most maintainable way to go.

xworld21 replied:
Old dependency resolution moved to `combine-wrapper.nix`.

xworld21 commented on pkgs/tools/typesetting/tex/texlive/default.nix:
```diff
@@ -518,7 +536,7 @@ in
     combine = assert assertions; combine;
 
-    # Pre-defined combined packages for TeX Live schemes,
+    # Pre-defined evironments packages for TeX Live schemes,
```
```suggestion
    # Pre-defined evironment packages for TeX Live schemes,
```

xworld21 commented on pkgs/tools/typesetting/tex/texlive/default.nix:
```diff
@@ -502,8 +502,26 @@ let
+      pkgs = p.pkgs;
+    }) attrs.__pkgs // attrs;
+
+  __buildEnv = import ./buildenv.nix {
+    self = { inherit bin; __pkgs = tl; };
+    ghostscript = ghostscript_headless;
+    inherit lib buildEnv libfaketime makeFontsConf makeWrapper runCommand
+      writeText bash perl coreutils gawk gnugrep gnused;
+  };
```
I forgot that `buildEnv` is a *derivation* in Lua, Perl, etc. I presume because it splices correcly.

xworld21 replied:
It's now a derivation on which you call `.override`.

xworld21 replied:
I understand now why `buildEnv` is a derivation in Lua etc, and it has to do with how the packages are exported at the top level (`luaPackages`, `lua54Packages`, `lua5_4` etc). Since `texlive` is a recurse-for-derivation set, `buildEnv` must be a function or it will pop up in search, which of course we do not want. This can change later if we go for `texlive2022`, `texlive2022Full`, `texlivePackages` etc.

xworld21 commented:
Apologies for the confusion: I was rushing things through as I'll have little time over the next few weeks. We should review each group of commits on its own, in this case starting first with #250963. The draft of `__buildEnv` here serves to verify that the multi-output implementation has the functionality required. In fact I'll now update my scope draft too, again to verify that multi-output works as intended.

xworld21 commented on pkgs/tools/typesetting/tex/texlive/default.nix:
```diff
@@ -10,6 +10,9 @@
 , libfaketime, asymptote, biber-ms, makeFontsConf
 , useFixedHashes ? true
 , recurseIntoAttrs
+, extraName ? "combined-infraonly"
+, extraOutputsToInstall ? [ "man" ]
+, requiredTeXPackages ? (ps: [ ps.scheme-infraonly ])
```
I have opted for using what we already have:
```nix
texlive.override { extraOutputsToInstall = [ "man" "texdoc" ]; }
texlive.withPackages (ps: [ ... ])
```
instead of exposing a new `buildEnv`, which would be something new to maintain. I think that this is where I would prefer to go.

The PR is still incomplete in that `texlive.combined.scheme-small.override` has the arguments of `__buildEnv` rather than coming from `callPackage`, and that would be weird in practice.

xworld21 replied:
Changed my mind drastically.

xworld21 committed:
texlive.buildTeXLivePackage: switch to fake multi-output derivations for TeX Live packages

xworld21 committed:
tests.texlive.binaries: use new texlive package source

xworld21 committed:
texlive.combine: only derivations under .packages

xworld21 committed:
texlive.tlpdb.nix: detect presence of info pages

xworld21 committed:
mftrace: use new tex package structure

xworld21 committed:
biber-ms: use new texlive package source

xworld21 committed:
texlive.bin.xpdfopen: use new texlive package source

xworld21 committed:
biber: use new texlive package source

xworld21 committed:
texlive.bin.pygmentex: use new texlive package source

xworld21 committed:
sagetex: use new tex package structure

xworld21 committed:
tests.texlive.shebangs: use new texlive package source

xworld21 committed:
texlive: document new texlive.pkgs attribute

xworld21 committed:
iwona: use new texlive package source

xworld21 commented on pkgs/tools/typesetting/tex/texlive/build-tex-env.nix:
```diff
@@ -87,14 +130,45 @@ let
+    inherit passthru;
+  } // { outputSpecified = true; }));
+
+  passthru = splitOutputs // {
+    all = builtins.attrValues splitOutputs;
+    outputs = [ "out" ] ++ pkgList.nonEnvOutputs;
+    # This is set primarily to help find-tarballs.nix to do its job
+    requiredTeXPackages = lib.filter lib.isDerivation pkgList.all;
+    # useful for inclusion in the `fonts.packages` nixos option or for use in devshells
+    fonts = "${texmfroot}/texmf-dist/fonts";
+    overrideTeXConfig = newArgs: self (args // { __calledFromCombine = false; } // newArgs);
+    withPackages = reqs: self (args // { requiredTeXPackages = ps: requiredTeXPackages ps ++ reqs ps; __calledFromCombine = false; });
+  };
```
I have decided that fake multi-output is better here, where all TeX-relevant outputs are captured in `texmfdist` and the likes, and the rest are combined together in separate outputs.

99% of the time, it will result in additional `man` and `info` outputs that can be installed on demand, but will be ignored when using texlive as build input. Why do that: too many things can happen when a user can pass a non-texlive package (e.g. what happens to its man pages?), but with this approach, the result is ver predictable.

Apart from that, the main feature here is `.overrideTeXConfig`. This is essentially equivalent to a normal `.override` from `makeOverridable`, but with a different name to distinguish it from `texlive.override` (which changes the inputs).

xworld21 commented:
Documentation is in: https://github.com/NixOS/nixpkgs/blob/4af5649562f760dcd7cb1cc50d649d43018f4233/doc/languages-frameworks/texlive.section.md

I believe this is ready for a proper review, and merge if people like it.

veprbl commented on pkgs/tools/typesetting/tex/texlive/build-tex-env.nix:
```diff
@@ -1,52 +1,108 @@
+, requiredTeXPackages ? ps: [ ps.scheme-infraonly ]
+
+### texlive.combine backward compatibility
+, __extraName ? "combined"
+, __extraVersion ? ""
+# emulate the old texlive.combine (e.g. add man pages to main output)
+, __combine ? false
+# adjust behavior further if called from the texlive.combine wrapper
+, __fromCombineWrapper ? false
```
What does the `__` convey here?

xworld21 replied:
I expect these arguments to disappear once `texlive.combine` is phased out completely, and they are only used internally to mimic the old behaviours. Re `__extraName`, I wouldn't commit on a particular interface, and after all I don't think it's really needed.

xworld21 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5427,6 +5427,7 @@ with pkgs;
 
   # TeX Live; see https://nixos.org/nixpkgs/manual/#sec-language-texlive
   texlive = recurseIntoAttrs (callPackage ../tools/typesetting/tex/texlive { });
+  inherit (texlive.schemes) texliveBasic texliveContext texliveFull texliveGust texliveInfraonly texliveMedium texliveMinimal texliveSmall texliveTetex;
```
Do you mean why `texliveBasic` instead of `texliveSchemeBasic`? I think I was just parroting `python3Full`.

Come to think of it, Debian uses `texlive-full`, Fedora uses `texlive-scheme-full`. Our combined yielded `texlive-combined-full`. So no particular reason, really. Although if we do multiple years, `texlive2023SchemeFull` gets long.

Sidenote, `name` is always `texlive-env-2022` (should be `texlive-2022-env` actually!). Requested packages get listed in `meta.longDescription`.

(I suppose we agree that top level schemes is better than the current `callPackage { texLive = texlive.combined.scheme-small; }` approach. I'd rather have packages work like `asymptote` in this PR.)

xworld21 commented on pkgs/tools/typesetting/tex/texlive/build-tex-env.nix:
```diff
@@ -1,55 +1,111 @@
+      if __fromCombineWrapper then requiredTeXPackages null else
+        builtins.catAttrs "p" (builtins.genericClosure {
+          startSet = map keySet runtime.right;
+          operator = s: map keySet s.tlDeps;
```
```suggestion
          operator = s: map keySet s.tlDeps or (s.requiredTeXPackages or (_: [ ]) tl);
```
forgot to actually implement `requiredTeXPackages`

xworld21 commented on pkgs/test/texlive/default.nix:
```diff
@@ -613,7 +613,7 @@ rec {
                 (acc: pkg: concatLicenses acc (lib.toList (pkg.meta.license or [])))
                 []
-                scheme.passthru.packages;
+                scheme.passthru.requiredTeXPackages;
```
```suggestion
```
TODO: check if `recurseIntoAttrs` on `texlive.pkgs` (better yet `texlivePackages`) makes `find-tarballs.nix` work out of the box, in which case we can delete this line.

xworld21 replied:
To be clear: this should not be implemented here. Adding `recurseIntoAttrs` on `texlivePackages` would surface all TeX Live packages in search, thus requiring appropriate `meta.description` and warnings that the packages are for `withPackages`. And maybe there are less obtrusive ways to trick `find-tarballs.nix`.

xworld21 commented on pkgs/tools/typesetting/tex/texlive/build-tex-env.nix:
```diff
@@ -1,55 +1,111 @@
+    all =
+      let
+        # order of packages is irrelevant
+        packages = builtins.sort (a: b: a.pname < b.pname) (requiredTeXPackages tl);
```
While adding upgrade examples to the documentation, I realised that dropping support for old style `passthru.pkgs` packages is too harsh. People should be allowed to upgrade piecemeal.

Solution: add a preprocessing step to `requiredTeXPackages tl` that checks if the `.pkgs` attribute is a list of packages with `.tlType`, and if so, convert them in the same way as in `combine-wrapper.nix`. Seems reliable enough.

xworld21 commented on doc/languages-frameworks/texlive.section.md:
```diff
@@ -1,42 +1,54 @@
-Since release 15.09 there is a new TeX Live packaging that lives entirely under attribute `texlive`.
+Since release 15.09 there is a new TeX Live packaging that lives entirely under attribute `texlive`, with a few prebuilt environments exported at the top level, such as `texliveSmall`, `texliveFull`.
+
+Since release 23.11, `texlive` uses `.withPackages` to create environments, in line with other languages. The previous `texlive.combine` function is deprecated and will be removed in future releases.
```
Add examples of how to translate `combine` to `withPackages`.

xworld21 committed:
texlive: implement __overrideTeXConfig and withPackage

Implement new 'buildTeXEnv' to leverage multi-output packages and provide
__overrideTeXConfig/withPackages for modifying the configuration or adding
packages. The override mechanism is prefixed until stabilized.

xworld21 commented:
Now with better documentation (including upgrade examples) https://github.com/NixOS/nixpkgs/blob/dd116774e4e9fe3a103c1284dc588ab34ada854e/doc/languages-frameworks/texlive.section.md
and virtually 100% backward compatibility (now `withPackages` can cope with old style `passthru.pkgs` TeX packages).

Note: we could start by merging the first commits only, to do more testing, then merge documentation & top level changes when ready. Writing the documentation was mainly an exercise to spot the remaining backward compatibility issues.

xworld21 commented:
pinging @apfelkuchen6, if you can review this soon, that would be great - I am running a bit late with this but we should still be in time for 23.11

apfelkuchen6 commented:
@xworld21 Sorry for being so inactive in the past few weeks, I've had a lot on my plate. 
I'll try to find some time tomorrow to take a look at thisk

xworld21 commented:
> @xworld21 Sorry for being so inactive in the past few weeks, I've had a lot on my plate.
> I'll try to find some time tomorrow to take a look at thisk

No worries! The truly important question is whether the interfaces `.overrideTeXConfig` and `.withPackages` sound good. If so, I suppose we can merge, and fix the bugs as they arrive (if @veprbl agrees with the plan of course!).

xworld21 commented:
@veprbl I have removed all the treewide & documentation changes, and prefixed `__overrideTeXConfig`. I'd really like to see at least `.withPackages` make it by 23.11, since its behaviour can be stabilized right away. We can keep discussing `__overrideTeXConfig` as we go along.

(We need more reviewers! The three of us are not enough to keep a reliable pace...)
